python/pyarrow/tests/interchange/test_extra.py

```diff
@@ -0,0 +1,221 @@
+    )
+
+    pandas_df = table.to_pandas()
+    result = from_dataframe(pandas_df)
+
+    # Checking equality for the values
+    # As the dtype of the indices is changed from int32 in pa.Table
+    # to int64 in pandas interchange protocol implementation
+    assert result[0].chunk(0).dictionary == table[0].chunk(0).dictionary
+
+    table_protocol = table.__dataframe__()
+    result_protocol = result.__dataframe__()
+
+    assert table_protocol.num_columns() == result_protocol.num_columns()
+    assert table_protocol.num_rows() == result_protocol.num_rows()
+    assert table_protocol.num_chunks() == result_protocol.num_chunks()
+    assert table_protocol.column_names() == result_protocol.column_names()
+
+    col_table = table_protocol.get_column(0)
+    col_result = result_protocol.get_column(0)
+
+    assert col_result.dtype[0] == DtypeKind.CATEGORICAL
+    assert col_result.dtype[0] == col_table.dtype[0]
+    assert col_result.size == col_table.size
+    assert col_result.offset == col_table.offset
+
+    desc_cat_table = col_result.describe_categorical
+    desc_cat_result = col_result.describe_categorical
+
+    assert desc_cat_table["is_ordered"] == desc_cat_result["is_ordered"]
+    assert desc_cat_table["is_dictionary"] == desc_cat_result["is_dictionary"]
+    assert isinstance(desc_cat_result["categories"]._col, pa.Array)
+
+
+@pytest.mark.pandas
+def test_pandas_roundtrip():
+    if Version(pd.__version__) < Version("1.5.0"):
+        pytest.skip("__dataframe__ added to pandas in 1.5.0")
+    from datetime import datetime as dt
+
+    table = pa.table(
+        {
+            "a": [1, 2, 3, 4],  # dtype kind INT = 0
+            "b": [3, 4, 5, 6],  # dtype kind INT = 0
+            "c": [1.5, 2.5, 3.5, 4.5],  # dtype kind FLOAT = 2
+            "d": [9, 10, 11, 12],  # dtype kind INT = 0
+            "e": [True, True, False, False],  # dtype kind BOOLEAN = 20
+            "f": ["a", "", "c", "d"],  # dtype kind STRING = 21
+            "g": [dt(2007, 7, 13), dt(2007, 7, 14),
+                  dt(2007, 7, 15), dt(2007, 7, 16)]  # dtype kind DATETIME = 22
+        }
+    )
+
+    pandas_df = table.to_pandas()
+    result = from_dataframe(pandas_df)
+    assert table.equals(result)
+
+    table_protocol = table.__dataframe__()
+    result_protocol = result.__dataframe__()
+
+    assert table_protocol.num_columns() == result_protocol.num_columns()
+    assert table_protocol.num_rows() == result_protocol.num_rows()
+    assert table_protocol.num_chunks() == result_protocol.num_chunks()
+    assert table_protocol.column_names() == result_protocol.column_names()
+
+
+@pytest.mark.pandas
+def test_large_string():
+    if Version(pd.__version__) < Version("1.5.0"):
+        pytest.skip("__dataframe__ added to pandas in 1.5.0")
+
+    from pandas.core.interchange.from_dataframe import (
+        from_dataframe as pandas_from_dataframe
+    )
+
+    data = np.array([b'x'*1024]*(3*1024**2), dtype='object')  # 3GB bytes data
```

Comment:
I think the [failing tests](https://github.com/apache/arrow/actions/runs/3682323841/jobs/6229859945) have something to do with the LargeStringArray. Will test it on the CI later with using smaller array.